tasks/resolve_reference.rb

```diff
@@ -32,77 +35,34 @@ def config_client(opts)
   end
 
-  def resolve_reference(opts)
-    client = config_client(opts)
+  def resolve_reference(config, template)
+    unless template.key?(:uri) || template.key?(:name)
+      msg = "You must provide a 'name' or 'uri' in 'target_mapping' for the AWS plugin"
+      raise TaskHelper::Error.new(msg, 'bolt-plugin/validation-error')
+    end
+
+    client = config_client(config)
     resource = Aws::EC2::Resource.new(client: client)
 
     # Retrieve a list of EC2 instances and create a list of targets
     # Note: It doesn't seem possible to filter stubbed responses...
-    resource.instances(filters: opts[:filters]).map do |instance|
-      next unless instance.state.name == 'running'
-      target = {}
+    targets = resource.instances(filters: config[:filters]).select {|i| i.state.name == 'running'}
 
-      if opts.key?(:uri)
-        uri = lookup(instance, opts[:uri])
-        target['uri'] = uri if uri
-      end
-      if opts.key?(:name)
-        real_name = lookup(instance, opts[:name])
-        target['name'] = real_name if real_name
-      end
-      if opts.key?(:config)
-        target['config'] = resolve_config(instance, opts[:config])
+    attributes = config_lookup(template)
+    target_data = targets.map do |target|
+      attributes.each_with_object({}) do |attr, acc|
+        attr = attr.first
+        acc[attr] = target.respond_to?(attr) ? target.send(attr) : nil
       end
-
-      target if target['uri'] || target['name']
-    end.compact
-  end
-
-  # Look for an instance attribute specified in the inventory file
-  def lookup(instance, attribute)
-    instance.data.respond_to?(attribute) ? instance.data[attribute] : nil
-  end
-
-  # Walk the "template" config mapping provided in the plugin config and
-  # replace all values with the corresponding value from the resource
-  # parameters.
-  def resolve_config(name, config_template)
-    walk_vals(config_template) do |value|
-      if value.is_a?(String)
-        lookup(name, value)
-      else
-        value
-      end
-    end
-  end
-
-  # Accepts a Data object and returns a copy with all hash and array values
-  # Arrays and hashes including the initial object are modified before
-  # their descendants are.
-  def walk_vals(data, skip_top = false, &block)
-    data = yield(data) unless skip_top
-    if data.is_a? Hash
-      map_vals(data) { |v| walk_vals(v, &block) }
-    elsif data.is_a? Array
-      data.map { |v| walk_vals(v, &block) }
-    else
-      data
     end
-  end
 
-  def map_vals(hash)
-    hash.each_with_object({}) do |(k, v), acc|
-      acc[k] = yield(v)
-    end
+    populate_mapping(template, target_data)
   end
 
-  def task(opts)
-    targets = resolve_reference(opts)
-    return { value: targets }
-  rescue TaskHelper::Error => e
-    # ruby_task_helper doesn't print errors under the _error key, so we have to
-    # handle that ourselves
-    return { _error: e.to_h }
+  def task(opts = {})
+    template = opts.delete(:target_mapping) || {}
```

Comment:
Is there a reason you split this out in the task method instead of resolve_reference? It seems like both are required there.

Reply:
It just made writing the spec tests a bit easier since I didn't have to include `target_mapping` in the parameters to `resolve_reference`. Also made munging those objects separately a bit more readable. I think that's worth the tradeoff for having the simpler `task` function

Reply:
I dont really follow. It seems like this adds (very slightly) more complexity to the `task` function. To me it seems like that logic belongs in `resolve_reference`. I guess fundamentally what does not make sense to me is why would we break apart opts, just to pass it all through?

Reply:
It makes this object (and manipulating the object) slightly more verbose in tests https://github.com/puppetlabs/puppetlabs-aws_inventory/pull/3/files#diff-7b9156ad5c46017f1c099ed24fa5c5f1R30 - you're totally right, I'll change it to be in `resolve_reference`. I was just being lazy about the testing

Reply:
Updated